From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sleigh-InSPECtor <sleighinspector@outlook.com>
Date: Fri, 24 May 2024 15:56:07 +0930
Subject: [PATCH] 6584: AArch32: vst2 had out-of-range invalid varnode access

AArch32: fixed vst2

* Varnode ptr increment/decrement ignored bounds of attachent memory space, causing out-of-range invalid varnode access.
---
 .../ARM/data/languages/ARMneon.sinc           | 28 ++++++-------------
 1 file changed, 8 insertions(+), 20 deletions(-)

diff --git a/Ghidra/Processors/ARM/data/languages/ARMneon.sinc b/Ghidra/Processors/ARM/data/languages/ARMneon.sinc
index e8aedc75d5..315f80ac94 100644
--- a/Ghidra/Processors/ARM/data/languages/ARMneon.sinc
+++ b/Ghidra/Processors/ARM/data/languages/ARMneon.sinc
@@ -5455,14 +5455,10 @@ thv_vst1DdElement2: Dd^"["^thv_vst1Index^"]"	is Dd & thv_vst1Index & thv_c1011=2
 # VST2 (multiple 2-element structures)
 #
 
-vst2Dd: Dreg		is Dreg & ((TMode=0 & c0607=0) | (TMode=1 & thv_c0607=0))  & regInc
+vst2Dd: Dreg		is Dreg & Dreg2 & ((TMode=0 & c0607=0) | (TMode=1 & thv_c0607=0))  & regInc
 {
 	ptr1:4 = &Dreg;
-@if ENDIAN == "little"
-  	ptr2:4 = &Dreg + (regInc * 8);
-@else # ENDIAN == "big"
-  	ptr2:4 = &Dreg - (regInc * 8);
-@endif # ENDIAN = "big"
+	ptr2:4 = &Dreg2;
 	mult_dat8 = 8;
 <loop>
 	*:1 mult_addr = *[register]:1 ptr1;
@@ -5476,14 +5472,10 @@ vst2Dd: Dreg		is Dreg & ((TMode=0 & c0607=0) | (TMode=1 & thv_c0607=0))  & regIn
 	goto <loop>;
 <loop_end>
 }
-vst2Dd: Dreg		is Dreg & ((TMode=0 & c0607=1) | (TMode=1 & thv_c0607=1))  & regInc
+vst2Dd: Dreg		is Dreg & Dreg2 & ((TMode=0 & c0607=1) | (TMode=1 & thv_c0607=1))  & regInc
 {
 	ptr1:4 = &Dreg;
-@if ENDIAN == "little"
-  	ptr2:4 = &Dreg + (regInc * 8);
-@else # ENDIAN == "big"
-  	ptr2:4 = &Dreg - (regInc * 8);
-@endif # ENDIAN = "big"
+	ptr2:4 = &Dreg2;
 	mult_dat8 = 4;
 <loop>
 	*:2 mult_addr = *[register]:2 ptr1;
@@ -5497,14 +5489,10 @@ vst2Dd: Dreg		is Dreg & ((TMode=0 & c0607=1) | (TMode=1 & thv_c0607=1))  & regIn
 	goto <loop>;
 <loop_end>	
 }
-vst2Dd: Dreg		is Dreg & ((TMode=0 & c0607=2) | (TMode=1 & thv_c0607=2)) & regInc
+vst2Dd: Dreg		is Dreg & Dreg2 & ((TMode=0 & c0607=2) | (TMode=1 & thv_c0607=2)) & regInc
 {
 	ptr1:4 = &Dreg;
-@if ENDIAN == "little"
-  	ptr2:4 = &Dreg + (regInc * 8);
-@else # ENDIAN == "big"
-  	ptr2:4 = &Dreg - (regInc * 8);
-@endif # ENDIAN = "big"
+	ptr2:4 = &Dreg2;
 	mult_dat8 = 2;
 <loop>
 	*:4 mult_addr = *[register]:4 ptr1;
@@ -5519,8 +5507,8 @@ vst2Dd: Dreg		is Dreg & ((TMode=0 & c0607=2) | (TMode=1 & thv_c0607=2)) & regInc
 <loop_end>	
 }
 
-buildVst2DdListA:							is counter=0								{ }
-buildVst2DdListA: vst2Dd,buildVst2DdListA	is vst2Dd & buildVst2DdListA & esize0607	[ counter=counter-1; regNum=regNum+1; ] 
+buildVst2DdListA:							is counter=0								[ reg2Num=reg2Num-counter2; ] { }
+buildVst2DdListA: vst2Dd,buildVst2DdListA	is vst2Dd & buildVst2DdListA & esize0607	[ counter=counter-1; regNum=regNum+1; reg2Num=reg2Num+1; ]
 {
 	build vst2Dd;
 	build buildVst2DdListA;
-- 
2.45.1

